Config: fix bootloader config errors - propagate errors

0xc0170 opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

See https://github.com/ARMmbed/mbed-os/pull/12007, this fixes the config failures. Do not ignore all config failures

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

Not needed

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [X] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [X] No Tests required for this change (E.g docs only update)
    [] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

Commit:
Remove hiding of config errors related to RAM and ROM regions handling.

Currently any misconfiguration of, for example, bootloader feature will cause the
build system to just silently drop it and continue building which can lead to
completed builds of something the user didn't want to build in worst case and
failing builds after compilation (=wasted time) in the best.

0xc0170 committed:
tools: regions only if bootloader is supported

We removed catching and passing, we want to know for any misconfiguration if bootloader
is supported. Regions should check if bootloader is supported. In case not, just return.
Otherwise we catch any error.

This should help us to uncover missing regions or other config error (in case bootloader
is enabled via bootloader_supported set to true).

0xc0170 commented on tools/toolchains/mbed_toolchain.py:
```diff
@@ -894,63 +894,56 @@ def _add_all_regions(self, region_list, active_region_name):
         """
+
+        if not getattr(self.target, "bootloader_supported", False):
+            return
```
This is the addition for  #12007 . The rest was approved already.

madchutney commented on tools/toolchains/mbed_toolchain.py:
```diff
@@ -894,63 +894,56 @@ def _add_all_regions(self, region_list, active_region_name):
         """Add regions to the build profile, if there are any.
         """
+
+        if not getattr(self.target, "bootloader_supported", False):
```
I would rather encode the default behaviour in `self.target` rather than define the default here, other places where this is used may assume a different default. There is also a danger of typos using this technique that would not cause exceptions as if there is a typos the result is always `False`. Maybe these are relatively minor points but they do somewhat hinder maintainability so should be avoided in code we wish to maintain in the long term.

0xc0170 replied:
This is how it is done in regions().
 It should be done properly but I don't have entire picture how config works yet . I found this the easiest way to fix now (same check is used inside config but in the later phase when its too late to go back) and later to even rewrite this add_regions to catch errors better.

Do you suggest to have target method to return if bootloader is supported and add_regions() would not even be called if not?

0xc0170 committed:
UNO_91: disable bootloader support as device name missing

The config failes with an exception that was previously ignored.

0xc0170 committed:
TB_SENSE_1: fix bootloader, not supported, ROM regions here

0xc0170 committed:
RZ_A1H: bootloader support disable

Until the error is fixed:
    "start not found in targets.json.".format(active_memory)

0xc0170 committed:
NUCLEO_WB55RG: disable bootloader support

device_name is missing

0xc0170 committed:
GR_LYCHEE: disable bootloader support

An error to be fixed:
tools.config.ConfigException: Missing a memory that is read, write in CMSIS Pack data

0xc0170 committed:
DISCO_L4R9I: disable bootloader

device_name is missing

0xc0170 committed:
CYW9P62S1_43438EVB_01: disable bootloader support

device_name missing

0xc0170 committed:
CY8CPROTO_063_BLE: disable bootloader support

device_name is missing

0xc0170 committed:
CY8CPROTO_062_4343W: disable bootloader support

device_name is missing

0xc0170 committed:
CY8CKIT_062_WIFI_BT: disable bootloader support

device_name is missing

0xc0170 committed:
CY8CKIT_062_BLE: disable bootloader support

 device_name is missing

0xc0170 committed:
UNO_91H: remove sectors, bootloader was disabled

This should be reverted once bootloader is supported.

0xc0170 committed:
STM32F756xG: fix IAR RAM size check

jeromecoutant commented on targets/TARGET_STM/TARGET_STM32F7/TARGET_STM32F756xG/device/TOOLCHAIN_IAR/stm32f756xg.icf:
```diff
@@ -26,7 +26,7 @@ ITCM-RAM:  16K (0x3FFF)   @0x00000000
 if (!isdefinedsymbol(MBED_ROM_SIZE)) { define symbol MBED_ROM_SIZE   = 0x200000; }
 if (!isdefinedsymbol(MBED_RAM_START)) { define symbol MBED_RAM_START  = 0x20010000; }
-if (!isdefinedsymbol(MBED_RAM_START)) { define symbol MBED_RAM_SIZE   = 0x40000; }
+if (!isdefinedsymbol(MBED_RAM_SIZE)) { define symbol MBED_RAM_SIZE   = 0x40000; }
```
+1

But as I am a copy paste expert... you need to update also stm32f746xg.icf, stm32f767xi.icf, and stm32f769xi.icf...

0xc0170 replied:
how come they don't fail the build? will check

0xc0170 replied:
Fixed now

0xc0170 replied:
Thanks a copy paste expert!

0xc0170 committed:
TARGET_STM32F74: fix IAR SIZE check in linker scripts
